includes/widgets/class-listings-widget.php

```diff
@@ -4,150 +4,196 @@
+		}
+
+		return $this->render_widget( $items, $instance, $html_class );
+	}
+
+	private function render_empty_widget( $html_class ) {
+		return sprintf( '<li class="wpbdp-empty-widget %s">%s</li>', $html_class, __( 'There are currently no listings to show.', 'business-directory-plugin' ) );
+	}
+
+	private function render_widget( $items, $instance, $html_class ) {
+		$html_class = implode(
+			' ',
+			array(
+				$this->get_item_thumbnail_position_css_class( $instance['thumbnail_desktop'], 'desktop' ),
+				$this->get_item_thumbnail_position_css_class( $instance['thumbnail_mobile'], 'mobile' ),
+				$html_class,
+			)
+		);
+
+		$show_images       = in_array( 'images', $this->supports ) && isset( $instance['show_images'] ) && $instance['show_images'];
+		$default_image     = $show_images && isset( $instance['default_image'] ) && $instance['default_image'];
+		$coming_soon_image = WPBDP_Listing_Display_Helper::get_coming_soon_image();
+		foreach ( $items as $post ) {
+			$html[] = $this->render_item( $post, compact( 'show_images', 'default_image', 'coming_soon_image', 'html_class' ) );
+		}
+
+		return join( "\n", $html );
+	}
+
+
+	private function get_item_thumbnail_position_css_class( $thumbnail_position, $version ) {
+		if ( $thumbnail_position == 'left' || $thumbnail_position == 'right' ) {
+			$css_class = sprintf( 'wpbdp-listings-widget-item-with-%s-thumbnail-in-%s', $thumbnail_position, $version );
+		} else {
+			$css_class = sprintf( 'wpbdp-listings-widget-item-with-thumbnail-above-in-%s', $version );
+		}
+
+		return $css_class;
+	}
+
+	private function render_item( $post, $args ) {
+		$output        = '';
+		$listing       = wpbdp_get_listing( $post->ID );
+		$listing_title = sprintf( '<div class="wpbdp-listing-title"><a class="listing-title" href="%s">%s</a></div>', $listing->get_permalink(), $listing->get_title() );
+		$html_image    = $this->render_image( $listing, $args );
+
+		if ( ! empty( $html_image ) ) {
+			$template = '<li class="wpbdp-listings-widget-item %1$s"><div class="wpbdp-listings-widget-container wpbdp-clearfix"><div class="wpbdp-listings-widget-thumb wpbdp-clearfix">%2$s</div><div class="wpbdp-listings-widget-item--title-and-content">%3$s</div></div></li>';
+		} else {
+			$args['html_class'] .= ' wpbdp-listings-widget-item-without-thumbnail';
+			$template            = '<li class="wpbdp-listings-widget-item %1$s"><div class="wpbdp-listings-widget-container wpbdp-clearfix"><div class="wpbdp-listings-widget-item--title-and-content">%3$s</div></div></li>';
+		}
+
+		return sprintf( $template, $args['html_class'], $html_image, $listing_title );
+	}
+
+	private function render_image( $listing, $args ) {
+		$image_link = '';
+		if ( $args['show_images'] ) {
+			$img_size = 'wpbdp-thumb';
+			if ( $img_id = $listing->get_thumbnail_id() ) {
+				$image_link = '<a href="' . $listing->get_permalink() . '">' . wp_get_attachment_image( $img_id, $img_size, false, array( 'class' => 'listing-image' ) ) . '</a>';
```

Comment:
Looks like there's a lot of escaping missing here too. Can you please review all the code in this class for escaping?

Reply:
Added escapes